### PR TITLE
Prepare `sc-network` for `ProtocolController`/`NotificationService`

### PR DESCRIPTION
**Pull request by altonen**
The upcoming notification protocol refactoring requires that protocols are able to communicate with `sc-network` over unique and direct links. This means that `sc-network` side of the link has to be created before `sc-network` is initialized and that it is allowed to consume the object as the receiver half of the link may not implement `Clone`.

Remove request-response and notification protocols from `NetworkConfiguration` and create a new object that contains the configurations of these protocols and which is consumable by `sc-network`. This is needed needed because, e.g., the receiver half of `NotificationService` is not clonable so `sc-network` must consume it when it's initializing the protocols in `Notifications`.

Similar principe applies to `PeerStore`/`ProtocolController`: as per current design, protocols are created before the network so `Protocol` cannot be the one creating the `PeerStore` object. `FullNetworkConfiguration` will be used to store the objects that `sc-network` will use to communicate with protocols and it will also allow protocols to allocate handles so they can directly communicate with `sc-network`.

polkadot companion: https://github.com/paritytech/polkadot/pull/7184
cumulus companion: https://github.com/paritytech/cumulus/pull/2526

### CODE REVIEW

File: client/network/sync/src/engine.rs [hunk 1]
```diff
@@ -272,52 +272,63 @@ where
...
+				let peer_ids = config
+					.set_config
+					.reserved_nodes
+					.iter()
+					.map(|info| info.peer_id)
+					.collect::<Vec<PeerId>>();
+				imp_p.extend(peer_ids);
 			}
+
+			// for reserved in network_config
```
Comment (melekes) [inline client/network/sync/src/engine.rs:311]: why is this commented out?

**Review by melekes — APPROVED** [review]
👍  minus one comment

File: client/network/sync/src/engine.rs (hunk 1 above)
Comment (altonen) [reply client/network/sync/src/engine.rs]: Fixed

**Review by skunert — APPROVED** [review]

**Review by dmitry-markin — APPROVED** [review]

**Comment by altonen** [comment]
bot merge

**Pull request merged**